docs/guide/api-environment-runtimes.md

```diff
@@ -328,36 +359,22 @@ export const runner = new ModuleRunner(
 await runner.import('/entry.js')
 ```
 
-## ModuleRunnerHMRConnection
-
-**Type Signature:**
+In this case, the server that handles those HTTP requests can use `environment.getInvokeHandlers()` to process those requests:
 
 ```ts
-export interface ModuleRunnerHMRConnection {
-  /**
-   * Checked before sending messages to the server.
-   */
-  isReady(): boolean
-  /**
-   * Send a message to the server.
-   */
-  send(payload: HotPayload): void
-  /**
-   * Configure how HMR is handled when this connection triggers an update.
-   * This method expects that the connection will start listening for HMR updates and call this callback when it's received.
-   */
-  onUpdate(callback: (payload: HotPayload) => void): void
-}
+server.onRequest((request: Request) => {
+  const payload = (await request.json()) as CustomPayload
+  if (payload.type === 'custom') {
+    const handler = invokeHandlers[payload.event]
```

Comment:
If we expose `handleInvokePayload` instead of `getInvokeHandlers`, doing things like this will be impossible.
https://github.com/vitejs/vite/blob/d499c9f31dd61829db600f5d9e8ac7d4a1f23929/packages/vite/src/node/server/environment.ts#L135-L143

Reply:
Probably I'm still confused in some places. First thing I want to confirm is that is it possible or does it make any sense for users to override `getInvokeHandler` etc... to make something other than `vite:fetchModule` invoke-able?

Reply:
I don't expect `getInvokeHandler` to be overridden by users.

Reply:
One idea I had is to wrap handlers into one and branch it inside if we need more events other than `vite:fetchModule`, something like:

```ts
// environment.ts
class DevEnvironment {
  constructor() {
    this.hot.on("vite:invoke", async (data, client, invoke) => {
      const result = await this.handleInvoke(data)
      client.respond('vite:invoke', invoke, result)
    })
  }

  handleInvoke(data) {
    if (data.method === 'fetchModule') {
      const result = await this.fetchModule(
        ...(data.args as [string, string | undefined, any]),
      )
      return { r: result }
    }
    ... other method
  }
}

// runner.ts
this.transport.invoke('vite:invoke', {
  method: 'fetchModule',
  args: [ ... ]
})
```

Reply:
I tried `transport.setInvokeHandler` in https://github.com/hi-ogawa/vite-environment-examples/pull/139/commits/a86f681fab56ee1d6f1a9dac824c6c1de18459e4, but it still feels cumbersome. Can we move `wrappedInvokeHandler` from `normalizeHotChannel` to `DevEnvironment.handleInvoke`?

I understand `invoke`-only transport is not much of our target, but I think it would worth it if we can still keep `fetchModule`-like simplicity.

Reply:
The more I play with this, it looks like this is just a minor convenience for minor use cases, so I don't think this is really a blocker. I brought it up, but feel free to ignore unless you have an elegant idea to organize.

Reply:
Providing `DevEnvironment.handleInvoke` would make `invoke`-only transport easier, but the downside of adding it is that it expands the API surface. I think environments have an incentive to provide as much functionality as possible, so probably they would implement in a way that can support `send` as well, even if it is a bit cumbersome. For `invoke`-only transport, I think it's better to reduce the API surface. If there's a way to make it `send` + `invoke` transport easier to implement, I think we should do it, but I don't have any ideas now.